src/MessageWriter.test.js

```diff
@@ -0,0 +1,413 @@
+      string url
+      uint8 id
+      `;
+
+      const writer = new MessageWriter(messageDefinition);
+      const buffer = Buffer.concat([
+        getStringBuffer("foo"),
+        // uint32LE length of Account array (2)
+        new Buffer([0x02, 0x00, 0x00, 0x00]),
+        // name
+        getStringBuffer("bar"),
+        // id
+        new Buffer([100, 0x00]),
+        // uint32LE length of Photo array (3)
+        new Buffer([0x03, 0x00, 0x00, 0x00]),
+        // photo url
+        getStringBuffer("http://foo.com"),
+        // photo id
+        new Buffer([10]),
+
+        // photo url
+        getStringBuffer("http://bar.com"),
+        // photo id
+        new Buffer([12]),
+
+        // photo url
+        getStringBuffer("http://zug.com"),
+        // photo id
+        new Buffer([16]),
+
+        // next account
+        getStringBuffer("baz"),
+        new Buffer([101, 0x00]),
+        // uint32LE length of Photo array (0)
+        new Buffer([0x00, 0x00, 0x00, 0x00]),
+      ]);
+
+      const message = {
+        username: "foo",
+        accounts: [
+          {
+            name: "bar",
+            id: 100,
+            photos: [
+              {
+                url: "http://foo.com",
+                id: 10,
+              },
+              {
+                url: "http://bar.com",
+                id: 12,
+              },
+              {
+                url: "http://zug.com",
+                id: 16,
+              },
+            ],
+          },
+          {
+            name: "baz",
+            id: 101,
+            photos: [],
+          },
+        ],
+      };
+
+      expect(writer.writeMessage(message)).toEqual(buffer);
+    });
+
+    const withBytesAndBools = `
+      byte OK=0
+      byte WARN=1
+      byte ERROR=2
+      byte STALE=3
+      byte FOO = 3 # the space exists in some topics
+      byte FLOAT64      = 8
+
+      bool level\t\t# level of operation enumerated above
+
+      DiagnosticStatus status
+
+      ================================================================================
+      MSG: diagnostic_msgs/DiagnosticStatus
+      # This message holds the status of an individual component of the robot.
+      #
+
+      # Possible levels of operations
+      byte OK=0
+      byte WARN=1  # Comment        # FLOATING OUT HERE
+      byte ERROR=2
+      byte STALE=3
+
+      byte    level # level of operation enumerated above
+      string name # a description of the test/component reporting`;
+
+    it("writes bytes and constants", () => {
+      const writer = new MessageWriter(withBytesAndBools);
+      const buffer = Buffer.concat([Buffer.from([0x01]), Buffer.from([0x00]), getStringBuffer("foo")]);
+
+      const message = {
+        level: true,
+        status: {
+          leve: 0,
+          name: "foo",
+        }
+      };
+
+      expect(writer.writeMessage(message)).toEqual(buffer);
+    });
+  });
+
+  describe("MessageReader and MessageWriter outputs are compatible", () => {
```

Comment:
?

Reply:
Sorry, missed this comment originally. I think it's valuable to have most tests focused on individual behavior of the reader/writer so that we can change and test those modules independently and aren't testing too much at once.

Reply:
np -- yea I figured the limited-scope might be a motivating factor. Sounds good.